### PR TITLE
Update/rewrite reddeer test suite run with requirements loaded from config file

### PR DESCRIPTION
**Pull request by odockal**
Aim is to be able to run multiple test runs on single test suite based on annotated requirements loaded from single config file. This is especially aiming to use of JRE and Server (only to those?) requirements together defined with one single config xml file. Such loaded requirement configuration would create matrix of combination of possible configurations that would test suite be run with.

### CODE REVIEW

**Comment by odockal** [comment]
DO NOT MERGE! 
This is first prove of concept commit. Its purpose is to try run multiple tests with more than one configuration used in config.xml file.

**Comment by odockal** [comment]
testPR

**Comment by rawagner** [comment]
We want to support only one config file which means that SuiteConfiguration.getListOfConfigurationsPerFile() should be renamed and should return just one file

**Comment by rawagner** [comment]
check support for nested suites

**Comment by rawagner** [comment]
require requirements .xml file not just folder

**Comment by mlabuda** [comment]
Could you please create more lightweight tests? Using server requirement and JRE requirement seems too much for me. Create 2 simple requirements (or if there are existing for testing, use those) and assert numbers of executed classes/tests for various combination are equal to expected numbers..

**Comment by odockal** [comment]
testPR

**Comment by rawagner** [comment]
test case:


@Server1
class Parent


@Server2
class Child extends Parent



Child class has @InjectRequirement Server - which requirement is injected - Server1 or Server2 ? We need some additional property for @InjectRequirement to specify what should be injected

**Comment by odockal** [comment]
testPR

**Comment by odockal** [comment]
@rawagner @mlabuda 
It seems that in rasta's case, where we are injecting server requirement in child class, the child's requirement is used. If child class has no "overloading" requirement, the parent's one is used when injecting requirement in child's class.

**Comment by rawagner** [comment]
require rd.config to point to .xml not just folder

**Comment by odockal** [comment]
@rawagner @mlabuda @apodhrad @rhopp @jkopriva @jrichter1 @theJNOVAK 
Take a look, amigos. Builds'n'tests seems be almost stable (in case of tests), PR still requires some work but I suggest to do it in the form of another commit/fixup.

File: plugins/org.jboss.reddeer.junit/src/org/jboss/reddeer/junit/internal/configuration/RequirementConfigurationObject.java [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+import org.jboss.reddeer.junit.requirement.PropertyConfiguration;
+import org.jboss.reddeer.junit.requirement.Requirement;
+import org.jboss.reddeer.junit.requirement.RequirementException;
+
+/**
+ * Represents requirement configuration class and its configuration object defined in xml file.
+ * @author odockal
+ *
+ */
+public class RequirementConfigurationObject {
```
Comment (rawagner) [inline plugins/org.jboss.reddeer.junit/src/org/jboss/reddeer/junit/internal/configuration/RequirementConfigurationObject.java]: rename to RequirementConfiguration

File: plugins/org.jboss.reddeer.junit/src/org/jboss/reddeer/junit/internal/configuration/RequirementsConfigurationImpl.java [hunk 2]
```diff
@@ -31,7 +32,7 @@
  * @author Lucia Jelinkova
  * @see TestRunConfiguration
  */
-public class RequirementsConfigurationImpl implements RequirementsConfiguration{
```
Comment (rawagner) [inline plugins/org.jboss.reddeer.junit/src/org/jboss/reddeer/junit/internal/configuration/RequirementsConfigurationImpl.java]: rename to RequirementConfiguratorImpl

File: plugins/org.jboss.reddeer.junit/src/org/jboss/reddeer/junit/internal/configuration/TestClassRequirementSet.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+	
+	public List<Class<?>> getClasses() {
+		return this.testClasses;
+	}
+	
+	/**
+	 * Gets all test classes as an array.
+	 * @return all test classes as an array
+	 */
+	public Class<?>[] getClassesAsArray() {
```
Comment (rawagner) [inline plugins/org.jboss.reddeer.junit/src/org/jboss/reddeer/junit/internal/configuration/TestClassRequirementSet.java]: one liner testClasses.toArray(new Class<?>[testClasses.size()]);

File: plugins/org.jboss.reddeer.junit/src/org/jboss/reddeer/junit/internal/configuration/RequirementConfigurationObject.java [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+	 * insert them into instance's list of {@link configurations}. 
+	 * @param reader the reader
+	 */
+	private void readConfigurationFromXML(XMLReader reader) {
+		
+		@SuppressWarnings("unchecked")
+		Requirement<Annotation> requirement = (Requirement<Annotation>) createInstance(this.requirementConfigurationClass.getEnclosingClass());
+		
+		if (!(requirement instanceof CustomConfiguration || requirement instanceof PropertyConfiguration)){
+			throw new IllegalArgumentException("The requirement does not implement " + CustomConfiguration.class);
```
Comment (rawagner) [inline plugins/org.jboss.reddeer.junit/src/org/jboss/reddeer/junit/internal/configuration/RequirementConfigurationObject.java]: Shouldnt you mention PropertyConfiguration.class too ?

File: plugins/org.jboss.reddeer.junit/src/org/jboss/reddeer/junit/internal/configuration/configurator/CustomConfigurator.java [hunk 5]
```diff
@@ -57,16 +58,23 @@ public void configure(Requirement<?> requirement) {
 		
 		log.debug("Configuration object associated with requirement " + requirement.getClass() + " is " + customConfiguration.getConfigurationClass());
 		
-		List<Object> configs = reader.getConfiguration(customConfiguration.getConfigurationClass());
-		if (configs.isEmpty()){
-			throw new RedDeerConfigurationException("No configuration found in XML configuration file for requirement class " + customConfiguration.getClass());
+		// iterate over all configurations and tries to cast the configuration object to expected configuration class
+		// Class cast exception is caught and iteration continues, if none configuration fits, new
```
Comment (rawagner) [inline plugins/org.jboss.reddeer.junit/src/org/jboss/reddeer/junit/internal/configuration/configurator/CustomConfigurator.java]: "new" what ? Something is missing in the comment

File: plugins/org.jboss.reddeer.junit/src/org/jboss/reddeer/junit/internal/configuration/configurator/CustomConfigurator.java [hunk 6]
```diff
@@ -57,16 +58,23 @@ public void configure(Requirement<?> requirement) {
...
 		log.debug("Configuration object associated with requirement " + requirement.getClass() + " is " + customConfiguration.getConfigurationClass());
 		
-		List<Object> configs = reader.getConfiguration(customConfiguration.getConfigurationClass());
-		if (configs.isEmpty()){
-			throw new RedDeerConfigurationException("No configuration found in XML configuration file for requirement class " + customConfiguration.getClass());
+		// iterate over all configurations and tries to cast the configuration object to expected configuration class
+		// Class cast exception is caught and iteration continues, if none configuration fits, new 
+		boolean configurationSet = false;
+		for (Object configuration : this.configurations) {
+			try {
```
Comment (rawagner) [inline plugins/org.jboss.reddeer.junit/src/org/jboss/reddeer/junit/internal/configuration/configurator/CustomConfigurator.java]: change try-catch to isAssignableFrom

File: plugins/org.jboss.reddeer.junit/src/org/jboss/reddeer/junit/internal/configuration/configurator/PropertyBasedConfigurator.java [hunk 7]
```diff
@@ -26,75 +23,60 @@
...
-		List<PropertyBasedConfiguration> list = reader.getConfiguration(PropertyBasedConfiguration.class);
-		Map<String, PropertyBasedConfiguration> map = new HashMap<String, PropertyBasedConfiguration>();
-		
-		for (PropertyBasedConfiguration config : list){
-			if (map.containsKey(config.getRequirementClassName())){
-				throw new RedDeerConfigurationException("There is more than one configuration in the XML file for requirement class " + config.getRequirementClassName());
+
+		boolean configurationSet = false;
+		for (Object configuration : this.configurations) {
+			try {
```
Comment (rawagner) [inline plugins/org.jboss.reddeer.junit/src/org/jboss/reddeer/junit/internal/configuration/configurator/PropertyBasedConfigurator.java]: change try-catch to isAssignableFrom

File: plugins/org.jboss.reddeer.junit/src/org/jboss/reddeer/junit/internal/runner/TestsExecutionManager.java [hunk 8]
```diff
@@ -51,6 +51,7 @@ public int compare(Class<?> clazz0, Class<?> clazz1) {
 	 * 				as test class without a run
 	 */
 	public void addTest(Class<?> testClass) {
+		System.out.println("Adding class without a run: " + testClass.getName());
```
Comment (rawagner) [inline plugins/org.jboss.reddeer.junit/src/org/jboss/reddeer/junit/internal/runner/TestsExecutionManager.java]: use logger

File: plugins/org.jboss.reddeer.junit/src/org/jboss/reddeer/junit/internal/runner/TestsExecutionManager.java [hunk 9]
```diff
@@ -62,6 +63,7 @@ public void addTest(Class<?> testClass) {
 	 * 				as test class with a run
 	 */
 	public void addExecutedTest(Class<?> testClass) {
+		System.out.println("Adding class with a run: " + testClass.getName());
```
Comment (rawagner) [inline plugins/org.jboss.reddeer.junit/src/org/jboss/reddeer/junit/internal/runner/TestsExecutionManager.java]: logger

File: plugins/org.jboss.reddeer.requirements/resources/v1/ServerRequirementsBase.xsd [hunk 10]
```diff
@@ -5,7 +5,7 @@
 
 	<!-- Import basic RedDeer requirements -->
 	<xs:import namespace="http://www.jboss.org/NS/Req"
-		schemaLocation="http://www.jboss.org/schema/reddeer/v1/RedDeerSchema.xsd" />
+		schemaLocation="RedDeerSchema2.xsd" />
```
Comment (rawagner) [inline plugins/org.jboss.reddeer.requirements/resources/v1/ServerRequirementsBase.xsd]: undo the change

File: plugins/org.jboss.reddeer.requirements/src/org/jboss/reddeer/requirements/jre/JRERequirement.java [hunk 11]
```diff
@@ -85,6 +86,12 @@ public boolean canFulfill() {
 		}
 	}
 
+	@Override
+	public boolean isDeclarationAcceptable() {
+		// TODO Delete random generating
```
Comment (rawagner) [inline plugins/org.jboss.reddeer.requirements/src/org/jboss/reddeer/requirements/jre/JRERequirement.java]: remove TODO

File: tests/org.jboss.reddeer.junit.test/src/org/jboss/reddeer/junit/test/integration/runner/order/TestSequenceRedDeerSuite.java [hunk 12]
```diff
@@ -53,17 +53,17 @@ public void run(RunNotifier notifier) {
 	private Throwable createException() {
 		Iterator<?> realSequenceIterator = TestSequence.getRealSequence().iterator();
 		Iterator<?> expectedSequenceIterator = getExpectedSequence().iterator();
-
+		
 		int i = 0;
 		while (realSequenceIterator.hasNext() && expectedSequenceIterator.hasNext()){
 			Object real = realSequenceIterator.next();
 			Object expected = expectedSequenceIterator.next();
+			System.out.println(real.toString() + " vs. " + expected.toString());
```
Comment (rawagner) [inline tests/org.jboss.reddeer.junit.test/src/org/jboss/reddeer/junit/test/integration/runner/order/TestSequenceRedDeerSuite.java]: log

File: plugins/org.jboss.reddeer.requirements/src/org/jboss/reddeer/requirements/jre/JRERequirement.java (hunk 11 above)
Comment (odockal) [inline plugins/org.jboss.reddeer.requirements/src/org/jboss/reddeer/requirements/jre/JRERequirement.java]: I deleted whole method. It is declared with default value in the interface.

File: tests/org.jboss.reddeer.junit.test/src/org/jboss/reddeer/junit/test/integration/runner/order/TestSequenceRedDeerSuite.java (hunk 12 above)
Comment (odockal) [inline tests/org.jboss.reddeer.junit.test/src/org/jboss/reddeer/junit/test/integration/runner/order/TestSequenceRedDeerSuite.java]: no need for output...

File: plugins/org.jboss.reddeer.junit/src/org/jboss/reddeer/junit/internal/configuration/RequirementConfigurationObject.java (hunk 4 above)
Comment (odockal) [inline plugins/org.jboss.reddeer.junit/src/org/jboss/reddeer/junit/internal/configuration/RequirementConfigurationObject.java]: Mentioned.

File: plugins/org.jboss.reddeer.junit/src/org/jboss/reddeer/junit/internal/configuration/configurator/CustomConfigurator.java (hunk 5 above)
Comment (odockal) [inline plugins/org.jboss.reddeer.junit/src/org/jboss/reddeer/junit/internal/configuration/configurator/CustomConfigurator.java]: Comment deleted, Method's javadoc description updated.

File: plugins/org.jboss.reddeer.junit/src/org/jboss/reddeer/junit/internal/configuration/configurator/PropertyBasedConfigurator.java (hunk 7 above)
Comment (odockal) [inline plugins/org.jboss.reddeer.junit/src/org/jboss/reddeer/junit/internal/configuration/configurator/PropertyBasedConfigurator.java]: One catch removed - added if with PropertyBasedConfiguration.class.isAssignableFrom(configuration.getClass()) condition.

File: plugins/org.jboss.reddeer.junit/src/org/jboss/reddeer/junit/internal/runner/TestsExecutionManager.java (hunk 8 above)
Comment (odockal) [inline plugins/org.jboss.reddeer.junit/src/org/jboss/reddeer/junit/internal/runner/TestsExecutionManager.java]: println deleted

File: plugins/org.jboss.reddeer.junit/src/org/jboss/reddeer/junit/internal/runner/TestsExecutionManager.java (hunk 9 above)
Comment (odockal) [inline plugins/org.jboss.reddeer.junit/src/org/jboss/reddeer/junit/internal/runner/TestsExecutionManager.java]: no need for output.

File: plugins/org.jboss.reddeer.requirements/resources/v1/ServerRequirementsBase.xsd (hunk 10 above)
Comment (odockal) [inline plugins/org.jboss.reddeer.requirements/resources/v1/ServerRequirementsBase.xsd]: undone.

File: plugins/org.jboss.reddeer.junit/src/org/jboss/reddeer/junit/internal/configuration/RequirementsConfigurationImpl.java (hunk 2 above)
Comment (odockal) [inline plugins/org.jboss.reddeer.junit/src/org/jboss/reddeer/junit/internal/configuration/RequirementsConfigurationImpl.java]: I left this classes intact, we agreed on that, right?

**Comment by odockal** [comment]
I need to update some unit tests for CustomConfigurator.

**Comment by rawagner** [comment]
testPR

**Comment by odockal** [comment]
@rawagner commits squashed.
